File: swift/Grid Analyzer/Grid Analyzer/Models/Domain/Coin.swift
//
//  Coin.swift
//  Grid Analyzer
//
//  Domain model representing a cryptocurrency
//

import Foundation

struct Coin: Identifiable, Equatable {
    let id = UUID()
    let symbol: String
    let prices: [Double]
    let minPrice: Double
    let maxPrice: Double
    let avgPrice: Double
    let changePercent: Double
    
    init(symbol: String, prices: [Double]) {
        self.symbol = symbol
        self.prices = prices
        self.minPrice = prices.min() ?? 0
        self.maxPrice = prices.max() ?? 0
        self.avgPrice = prices.isEmpty ? 0 : prices.reduce(0, +) / Double(prices.count)
        
        if let startPrice = prices.first,
           let endPrice = prices.last,
           startPrice > 0 {
            self.changePercent = ((endPrice - startPrice) / startPrice) * 100
        } else {
            self.changePercent = 0
        }
    }
    
    static func == (lhs: Coin, rhs: Coin) -> Bool {
        lhs.symbol == rhs.symbol
    }
} 

File: python/models/coin.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import List


@dataclass
class Coin:
	symbol: str
	prices: List[float]
	min_price: float = field(init=False)
	max_price: float = field(init=False)
	avg_price: float = field(init=False)
	change_percent: float = field(init=False)

	def __post_init__(self) -> None:
		self.min_price = min(self.prices) if self.prices else 0.0
		self.max_price = max(self.prices) if self.prices else 0.0
		self.avg_price = sum(self.prices) / len(self.prices) if self.prices else 0.0
		if self.prices and self.prices[0] != 0:
			self.change_percent = ((self.prices[-1] - self.prices[0]) / self.prices[0]) * 100.0
		else:
			self.change_percent = 0.0